scripts/enforce_style/main.py

```diff
@@ -0,0 +1,136 @@
+        r"^(([A-Z][a-z0-9_]*)+|[a-z][a-z0-9_]*)$")
+    type_style = re.compile(r"^([A-Z][a-z0-9_]*)+$")
+    any = re.compile(r"^.+$")
+    operators = ["=", "==", "!=", "()", "^", "^=", "&", "&=", "|", "|=", "<", "<=", ">", ">=", ">>", "<<", "[]",
+                 "+", "+=", "-", "-=", "*", "*=", "++", "--", "/", "/=", "->", "<<=", ">>=", "~", "!", "%", "%="]
+
+    allowed_types = ["int", "uint8_t", "uint16_t", "uint32_t", "uint64_t",
+                     "int8_t", "int16_t", "int32_t", "int64_t",
+                     "char", "bool", "void", "double", "float"]
+
+    incorrect_type_names = []
+    incorrect_function_names = []
+
+    def ValidateType(self, type_node):
+        name = type_node.spelling
+        name = name.replace("&", "").replace(
+            "*", "").replace("const", "").replace("typename", "").strip()
+
+        if name in self.allowed_types:
+            return True
+
+        if name.startswith("std::"):
+            return True
+
+        if name.startswith("asio::"):
+            return True
+
+        # Stripping templates
+        if name.endswith(">") and "<" in name:
+            name, _ = name.split("<", 1)
+
+        # Stripping namespace - must be done
+        # after stripping templates
+        if "::" in name:
+            _, name = name.rsplit("::", 1)
+
+        match = self.type_style.match(name)
+
+        if match is None:
+            print("Failed using", name)
+            self.incorrect_type_names.append(name)
+            return False
+
+        return True
+
+    def CheckPublicMemberFunction(self, node):
+        if "operator" in node.spelling:
+            op = node.spelling.replace("operator", "")
+            if op in self.operators:
+                return
+
+        if self.member_function_style.match(node.spelling) is None:
+            self.Error(node, "Function name does not follow style guide.")
+
+    def CheckPrivateMemberVariable(self, node):
+        if not node.spelling.endswith("_"):
+            self.Error(node, "Private member variables must end with '_'")
+
+    def CheckArgument(self, node):
+        return
+
+        # TODO: Does not work as clang rearranges with const first
+        if node.type.spelling.startswith("const "):
+            self.Error(node, "Incorrect placement of keyword 'const'")
+
+    def CheckReturnType(self, node):
+        if not self.ValidateType(node.result_type):
+            self.Error(node, "Return type name '" +
+                       node.result_type.spelling+"' does not follow style guide.")
+
+    def CheckClass(self, node):
+        pass
+
+    def CheckTypedef(self, node):
+        self.Error(node, "typedefs not permitted.")
+
+    def CheckUsingTypeAlias(self, node):
+        if not self.ValidateType(node):
+            self.Error(node, "Type alias '"+node.spelling +
+                       "' does not follow style guide.")
+
+
+if __name__ == "__main__":
+    ignores = ["vm/tokeniser.hpp"]
+
+    includes = sys.argv[2:]
```

Comment:
Please document what the args to this script should be, or an example of intended usage. Better yet, upgrade to argparse